### ISSUE TITLE
crash on combination of type generic, entity instance inside generate case string

### ISSUE THREAD

**Issue by jeras** [ghdl/ghdl#2593]
This was tested on latest Git master.

I tried the following shorter examples **not** causing a crash:
- generate case string + processes to print MODE,
- generate case string + entity instance without type generic
- generate case string + type generic, but no entity instances
- just entity instance without type generic.

As far as I tested I needed all 3 to cause a crash:
- type generic,
- entity instances,
- generate case string.

Here is the code:
```VHDL
library ieee;
use ieee.std_logic_1164.all;

entity mode_rtl is
   generic (
      type DAT_G
   );
   port (
      rxd_i : in  DAT_G;
      txd_o : out DAT_G
   );
end mode_rtl;

architecture rtl of mode_rtl is
begin
   txd_o <= rxd_i;
end rtl;


use std.env.all;

library ieee;
use ieee.std_logic_1164.all;

entity mode_tb is
   generic (
      MODE : string := "mode_x"
   );
end mode_tb;

architecture behavior of mode_tb is
   subtype DAT_G is std_logic_vector(8-1 downto 0);
   signal txd : DAT_G;
   signal rxd : DAT_G;
begin
   -- all MODE options must be of the same length to avoid tool errors
   gen_dut: case MODE generate
      when "mode_0" =>

         dut : entity work.mode_rtl
         generic map (
            DAT_G => DAT_G
         )
         port map (
            rxd_i => txd,
            txd_o => rxd
         );

      when others =>

         -- passthrough
         rxd <= txd;

   end generate gen_dut;

   p_test : process
   begin
      report "Simulation timeout." severity ERROR;
      finish;
   end process p_test;

end behavior;
```

**Comment by jeras** [comment]
Thanks for looking into it.

The example is still crashing if I change the default `MODE` generic to `mode_0` (which is a valid item in the `case` statement). Another way to cause a crash is to override the generic in the command line:
```sh
ghdl -r --std=08 work.mode_tb -gMODE="mode_0"
```

**Comment by jeras** [comment]
In the code this testcase was derived from I now get
```sh
hdl:error: NULL access dereferenced
```
I will work on this again in a couple of days, probably open a new issue.

**Comment by jeras** [comment]
I am not sure I will continue debugging this issue, so I will just dump some thoughts and experiences I have with passing generics/parameters/arguments applied at a command line.

I looked at GHDL documentation and there is a paragraph regarding generic choices being done at simulation time and this might change to only support passing generic overrides at elaboration time.

In Verilog there are 3 generalization methods available:
- macros (parser),
- parameters (elaboration),
- plusargs (simulation, runtime).
Each one is used at a different stage.

I used plusargs instead of parameters for RISC-V compliance tests, to pass file paths for testcase binary files and log files. Using Verilator this allowed me to compile the testbench once and then run many tests in parallel using make. There were no conflicts between files generated by simulations, therefore I was able to run tests in parallel.

 VHDL-2008 only has generics. VHDL-2019 provides access to system environment variables. If generic overrides are moved to elaboration, access to environment variables could be used in a similar fashion as Verilog plusargs.

I have an issue with the Vivado simulator, where I wrote a VHDL custom protocol wrapper around proprietary gigabit transceiver IP. It is supposed to support different lane numbers and different devices using generics. It works during synthesis, but the simulator complains some of the code is not compatible with the selected device even if it is not enabled with current generic choices.
